File: demo/csharp/lunar_eclipse/lunar_eclipse.cs
using System;
using demo_helper;
using CosineKitty;

namespace lunar_eclipse
{
    class Program
    {
        static int Main(string[] args)
        {
            AstroTime time;
            switch (args.Length)
            {
                case 0:
                    time = new AstroTime(DateTime.Now);
                    break;

                case 1:
                    time = DemoHelper.ParseTime("lunar_eclipse", args[0]);
                    break;

                default:
                    Console.WriteLine("USAGE: lunar_eclipse [date]");
                    return 1;
            }

            int count = 0;
            LunarEclipseInfo eclipse = Astronomy.SearchLunarEclipse(time);
            for(;;)
            {
                if (eclipse.kind != EclipseKind.Penumbral)
                {
                    PrintEclipse(eclipse);
                    if (++count == 10)
                        break;
                }
                eclipse = Astronomy.NextLunarEclipse(eclipse.center);
            }

            return 0;
        }

        static void PrintEclipse(LunarEclipseInfo eclipse)
        {
            // Calculate beginning/ending of different phases
            // of an eclipse by subtracting/adding the center time
            // with the number of minutes indicated by the "semi-duration"
            // fields sd_partial and sd_total.
            const double MINUTES_PER_DAY = 24 * 60;

            AstroTime p1 = eclipse.center.AddDays(-eclipse.sd_partial / MINUTES_PER_DAY);
            Console.WriteLine("{0}  Partial eclipse begins.", p1);

            if (eclipse.sd_total > 0.0)
            {
                AstroTime t1 = eclipse.center.AddDays(-eclipse.sd_total / MINUTES_PER_DAY);
                Console.WriteLine("{0}  Total eclipse begins.", t1);
            }

            Console.WriteLine("{0}  Peak of {1} eclipse.", eclipse.center, eclipse.kind.ToString().ToLowerInvariant());

            if (eclipse.sd_total > 0.0)
            {
                AstroTime t2 = eclipse.center.AddDays(+eclipse.sd_total / MINUTES_PER_DAY);
                Console.WriteLine("{0}  Total eclipse ends.", t2);
            }

            AstroTime p2 = eclipse.center.AddDays(+eclipse.sd_partial / MINUTES_PER_DAY);
            Console.WriteLine("{0}  Partial eclipse ends.", p2);
            Console.WriteLine();
        }
    }
}

File: demo/java/src/main/java/io/github/cosinekitty/astronomy/demo/LunarEclipse.java
package io.github.cosinekitty.astronomy.demo;

import java.util.stream.Stream;

import io.github.cosinekitty.astronomy.*;

public class LunarEclipse {
    /**
     * Searches for the next 10 partial/total lunar eclipses.
     *
     * @param startTime
     * The date and time after which to start searching for lunar eclipses.
     */
    public static int run(Time startTime) {
        LunarEclipseInfo e = Astronomy.searchLunarEclipse(startTime);
        Stream.iterate(e, x -> Astronomy.nextLunarEclipse(x.getPeak()))
                .filter(x -> x.getKind() != EclipseKind.Penumbral)
                .limit(10)
                .forEach(LunarEclipse::printEclipse);
        return 0;
    }

    private static void printEclipse(LunarEclipseInfo e) {
        // Calculate beginning/ending of different phases
        // of an eclipse by subtracting/adding the peak time
        // with the number of minutes indicated by the "semi-duration"
        // fields sdPartial and sdTotal.

        Time peak = e.getPeak();            // the central time of the eclipse, when the Moon is darkest
        double total = e.getSdTotal();      // the semiduration of totality in minutes, or 0 if not total
        double partial = e.getSdPartial();  // the semiduration of partiality in minutes
        Time p1 = peak.addDays(-e.getSdPartial() / Astronomy.MINUTES_PER_DAY);
        System.out.printf("%s  Partial eclipse begins.%n", p1);
        if (total > 0.0) {
            Time t1 = peak.addDays(-total / Astronomy.MINUTES_PER_DAY);
            System.out.printf("%s  Total eclipse begins.%n", t1);
        }
        System.out.printf("%s  Peak of %s eclipse.%n", peak, e.getKind().toString().toLowerCase());
        if (total > 0.0) {
            Time t2 = peak.addDays(+total / Astronomy.MINUTES_PER_DAY);
            System.out.printf("%s  Total eclipse ends.%n", t2);
        }
        Time p2 = peak.addDays(+partial / Astronomy.MINUTES_PER_DAY);
        System.out.printf("%s  Partial eclipse ends.%n", p2);
        System.out.println();   // extra blank line to separate each lunar eclipse
    }
}
